site/learn/expert-advice/authentication/webauthn-explained.md

```diff
@@ -0,0 +1,478 @@
+            "rk": true,
+            "up": true
+          },
+          "maxCredentialCountInList": 100,
+          "maxCredentialIdLength": 32,
+          "transports": [
+            "internal"
+          ],
+          "algorithms": [
+            {
+              "type": "public-key",
+              "alg": -257
+            }
+          ],
+          "minPINLength": 4,
+          "firmwareVersion": 19042
+        }
+      },
+      "statusReports": [
+        {
+          "status": "FIDO_CERTIFIED_L1",
+          "effectiveDate": "2020-08-05",
+          "certificationDescriptor": "Windows Hello Hardware Authenticator",
+          "certificateNumber": "FIDO20020190418002",
+          "certificationPolicyVersion": "1.3.6",
+          "certificationRequirementsVersion": "1.1.0"
+        }
+      ],
+      "timeOfLastStatusChange": "2020-08-05"
+    }
+```
+
+There is a lot of structured metadata that you can use if desired.
+
+## User Presence vs User Verification
+
+Authenticators can be also used as a second factor of authentication, when paired with a username and password or other factor. An authentication ceremony gives you assurances about the user having access to the authenticator. If you are using an authenticator in this way, you want to distinguish between user presence versus user verification. 
+
+User presence is when someone is there behind the authenticator who gave permission, but you don't know who. All it proves is that you have access to the authenticator. An example of this is when you press a button on a Yubikey.
+
+User verification, on the other hand, is when the user provides proof that they are who they say they are. The authenticator could require a pin, known only to the user. It could be a Face ID or Touch ID challenge. In any case, verification allows you to tie a specific user to the authentication ceremony.
+
+Of the two types, user verification is more prevalent with the platform authenticators than cross platform authenticators. And user presence is more useful as a second factor. When not used for MFA, you will typically want user verification.
+
+The initiating server specifies whether to use user verification. The three choices for user verification are preferred, required, and discouraged:
+
+* Preferred means that user verification is preferred, but not required. 
+* Required means that user verification is required and the authentication will fail if it cannot be performed. 
+* Discouraged means that user verification is not preferred and should not be used if possible. 
+
+## Resident Keys
+
+Another WebAuthn feature is resident keys. With these, the user to website mapping is stored, not on the server side, but on the authenticator.
+
+This is available for cross-platform authenticators. With this feature, the server does not have to have prior knowledge of the user before they are authenticated. They can register and authenticate at the same time. This is less common in the B2C space.
```

Comment:
You are correct that discoverable credentials can be used without first identifying the user. The way that they work is a bit different than what you've described.
- The user-credential mapping should still be stored on the server to validate that the credential used to authenticate belongs to the user authenticating (that server-side mapping may be the _only_ thing that associates a credential with the user. I don't think the authenticator response contains the user ID supplied during registration)
- The difference is that the _credential itself_ is stored on the authenticator for a discoverable credential
- The feature is not limited to cross-platform authenticators, but it will be more common to use them there as a cookie or other storage mechanism can be used to indicate a platform credential that is available
- Not sure what you mean by "register and authenticate at the same time." Registration and authentication are still two separate ceremonies even with client discoverable credentials
- I'm not sure whether this is less common in the B2C space as there aren't many examples to go by

The key differentiator for discoverable credentials, as you've mentioned, is that the Relying Party does not need to identify the user _before_ the authentication ceremony. For non-discoverable credentials, the user must be identified in some way (e.g. secure cookie, username/email field, etc.) before the authentication ceremony in order to determine which credentials are owned by that account. The RP returns these registered credentials as part of the credential request options object in the `allowCredentials` field.

Discoverable credentials, on the other hand, are requested by the RP by leaving the `allowCredentials` list _empty_. If there is a discoverable credential available, it will be used to complete the authentication ceremony, and the authenticator response will indicate which credential was used. That credential must be mapped to the user on the RP server.

Bonus tidbit: when a credential is not client discoverable, it may not be stored in the authenticator _at all_. The authenticator gathers the info it needs to complete the authentication ceremony, serializes it, then encrypts it with a symmetric private key. The `credentialId` in the registration response _is_ that encrypted payload. During authentication ceremonies, the authenticator attempts to decrypt the provided `credentialId`(s). If it is able to decrypt the value, the authenticator knows that it generated the passkey and then uses it to complete the authentication.

There are nuances to discoverable credentials I am not familiar with because they haven't been fully implemented in FusionAuth yet. Our planned workflows do not require them, though they could streamline the bootstrap authentication workflow.

Reply:
Thanks, that was helpful.